File: caper/samples/ruby/calc0.rb
require './calc0_parser.rb'

class Scanner
    attr_accessor :input

    def initialize input = $stdin
        @input = input
    end
    
    def get
        begin
            c = @input.getc
            return :token_eof, 0 if c.nil? || c == ?\n
        end while c =~ /\s/

        case c
        when ?+
            return :token_Add, 0
        when ?-
            return :token_Sub, 0
        when ?*
            return :token_Mul, 0
        when ?/
            return :token_Div, 0
        else
            return :token_eof, 0 if c.nil?
            if c =~ /\d/
                n = 0
                begin
                    n *= 10
                    n += c.ord - ?0.ord
                    c = @input.getc
                end while !c.nil? && c =~ /\d/
                @input.ungetc c if !c.nil?
                return :token_Number, n
            end
        end
        fail format("bad input char '%s'(%d)\n", c, c.ord)
    end
end

class SemanticAction
    def syntax_error
    end

    def stack_overflow
    end
    
    def downcast v
        v
    end

    def upcast v
        v
    end
    
    def Identity x
        x
    end

    def MakeAdd(x, y)
        $stderr.printf "%d + %d\n", x, y
        x + y
    end

    def MakeSub(x, y)
        $stderr.printf "%d - %d\n", x, y
        x - y
    end

    def MakeMul(x, y)
        $stderr.printf "%d * %d\n", x, y
        x * y
    end

    def MakeDiv(x, y)
        $stderr.printf "%d / %d\n", x, y
        x / y
    end
end

s = Scanner.new
parser = Calc::Parser.new(SemanticAction.new)

$stdout.print ">"

begin
    token, v = s.get
    #p Calc::token_label(token)
end while !parser.post(token, v)

if v = parser.accept
    printf "accepted %d\n", v
else
    print "failed\n"
end

File: caper/samples/php/calc0.php
<?php

error_reporting(E_ALL);

$output = '';
$errors = '';

function user_error_handler($errno, $errmsg, $filename, $linenum, $vars)
{
    global $errors;

    switch ($errno) {
    case E_USER_ERROR:
        $errors .= "ERROR: ";
        break;

    case E_USER_WARNING:
        $errors .= "WARNING: ";
        break;

    case E_USER_NOTICE:
        $errors .= "NOTICE: ";
        break;
        
    default:
        break;
    }
    $errors .= $filename . "(" . $linenum . "): " . $errmsg . "\n";
}
set_error_handler('user_error_handler');

require_once('calc0_parser.php');

class Scanner
{
    public $input;

    function __construct($data)
    {
        $this->input = $data;
    }
    
    function getch()
    {
        if (strlen($this->input) > 0)
        {
            $ch = $this->input[0];
            $this->input = substr($this->input, 1);
            return $ch;
        }
        return NULL;
    }
    
    function ungetch($ch)
    {
        $this->input = $ch . $this->input;
    }

    function get(&$v)
    {
        do
        {
            $c = $this->getch();
            if ($c == NULL || $c == "\n") {
                return \calc\Token::token_eof;
            }
        } while (preg_match('/\s/', $c));

        switch ($c)
        {
        case '+':
            return \calc\Token::token_Add;
        case '-':
            return \calc\Token::token_Sub;
        case '*':
            return \calc\Token::token_Mul;
        case '/':
            return \calc\Token::token_Div;
        default:
            if ($c == NULL) {
                return \calc\Token::token_eof;
            }
            if (preg_match('/\d/', $c)) {
                $n = 0;
                do
                {
                    $n *= 10;
                    $n += ord($c) - ord('0');
                    $c = $this->getch();
                } while ($c != NULL && preg_match('/\d/', $c));
                if ($c != NULL) {
                    $this->ungetch($c);
                }
                $v = $n;
                return \calc\Token::token_Number;
            }
        }
        trigger_error(sprintf("bad input char '%s'(%d)\n", $c, ord($c)), E_USER_ERROR);
        return \calc\Token::token_eof;
    }
}

class SemanticAction
{
    function syntax_error()
    {
    }

    function stack_overflow()
    {
    }
    
    function downcast($y)
    {
        return $y;
    }

    function upcast($y)
    {
        return $y;
    }
    
    function Identity($x)
    {
        return $x;
    }

    function MakeAdd($x, $y)
    {
        global $output;
        $output .= sprintf("%d + %d\n", $x, $y);
        return $x + $y;
    }

    function MakeSub($x, $y)
    {
        global $output;
        $output .= sprintf("%d - %d\n", $x, $y);
        return $x - $y;
    }

    function MakeMul($x, $y)
    {
        global $output;
        $output .= sprintf("%d * %d\n", $x, $y);
        return $x * $y;
    }

    function MakeDiv($x, $y)
    {
        global $output;
        $output .= sprintf("%d / %d\n", $x, $y);
        return $x / $y;
    }
}

header('Content-Type: text/html');

if (isset($_POST['input'])) {
    $input = $_POST['input'];
    $s = new Scanner($input);
    $parser = new calc\Parser(new SemanticAction());

    do
    {
        $token = $s->get($v);
    } while (!$parser->post($token, $v));

    if ($parser->accept($v)) {
        $result = "accepted " . $v;
    } else {
        $result = "failed";
    }
}

?><html><head>
<title>calc0</title>
</head>
<body>
    <h1>calc0</h1>
    <p style="color: red; font-weight: bold;">
        <?php
            if (isset($errors) && strlen($errors) > 0) {
                echo "ERROR: " . nl2br(htmlspecialchars($errors));
            }
        ?>
    </p>
    <p>
        <?php
            if (isset($input)) {
                echo "Input: " . htmlspecialchars($input);
            }
        ?>
    </p>
    <p>
        <?php
            if (isset($output) && strlen($output) > 0) {
                echo "Output: <br />" . nl2br(htmlspecialchars($output));
            }
        ?>
    </p>
    <p>
        <?php
            if (isset($result)) {
                echo "Result: " . htmlspecialchars($result);
            }
        ?>
    </p>
    <form action="calc0.php" method="post">
        <p>Input: <input type="text" name="input" value="" /></p>
        <p><input type="submit" value="Send" /></p>
    </form>
</body></html>
